### PR TITLE
Add partitioned index

### PR DESCRIPTION
**Pull request by woodsaj**
Add a PartitionedIndex wrapper around the memoryIdx.
Partitioning the index can be toggled via a config flag. By default it is off.

### CODE REVIEW

**Comment by shanson7** [comment]
Looks interesting. Is the goal of this component to reduce lock contention? Can you share any benchmarks run?

File: idx/memory/memory.go [hunk 1]
```diff
@@ -87,6 +89,23 @@ func ConfigProcess() {
 	}
 }
 
+// interface implemented by both MemoryIdx and PartitionedMemoryIdx
+// this is needed to support unit tests.
+type MemoryIndex interface {
```
Comment (replay) [inline idx/memory/memory.go]: I think it seems a bit confusing that there's now a thing called `MemoryIndex` and another thing called `MemoryIdx`. Wouldn't it make sense to rename `MemoryIdx` into something like `UnpartitionedMemoryIndex` or `MonolithicMemoryIndex` or whatever else describes the different between it and the `PartitionedMemoryIndex`?

File: idx/memory/partitioned_idx.go [hunk 2]
```diff
@@ -0,0 +1,536 @@
...
+		idx.Partition[p] = NewMemoryIdx()
+	}
+	return idx
+}
+
+// Init initializes the index at startup and
+// blocks until the index is ready for use.
+func (p *PartitionedMemoryIdx) Init() error {
+	for _, m := range p.Partition {
+		//log.Infof("PartitionedMemoryIdx: initializing MemoryIdx for partition %d", part)
```
Comment (replay) [inline idx/memory/partitioned_idx.go]: commented code

**Comment by replay** [comment]
if we already partition by partition id, should we also partition by org id? in the case of multi tenant setups this would keep the `defById` map smaller and it seems like it should be easy to now just make the key something like `(orgId * 128) + partition` assuming the max number of partitions is `128`.

File: idx/memory/partitioned_idx.go [hunk 3]
```diff
@@ -0,0 +1,536 @@
...
+
+// UpdateArchive updates the archive information
+func (p *PartitionedMemoryIdx) UpdateArchive(archive idx.Archive) {
+	p.Partition[archive.Partition].UpdateArchive(archive)
+}
+
+// Get returns the archive for the requested id.
+func (p *PartitionedMemoryIdx) Get(key schema.MKey) (idx.Archive, bool) {
+	g, _ := errgroup.WithContext(context.Background())
+	resultChan := make(chan *idx.Archive, len(p.Partition))
```
Comment (replay) [inline idx/memory/partitioned_idx.go]: is there any scenario where a key would be in more than one index partition? if not, then a length of `1` should be sufficient. if the reason for `len(p.Partition)` is to not block the insertion into the chan:

further down i think it would make more sense to do something like:
```
select {
    case err := g.Wait():
        if err != nil { .... }
    case result := <-resultChan:
      ...
}
```
otherwise we might end up waiting for a memory index partition even though the result is already ready in the chan

File: idx/memory/partitioned_idx.go [hunk 4]
```diff
@@ -0,0 +1,536 @@
...
+		return idx.Archive{}, false
+	}
+
+	return *result, true
+}
+
+// GetPath returns the archives under the given path.
+func (p *PartitionedMemoryIdx) GetPath(orgId uint32, path string) []idx.Archive {
+	g, _ := errgroup.WithContext(context.Background())
+	result := make([][]idx.Archive, len(p.Partition))
```
Comment (replay) [inline idx/memory/partitioned_idx.go]: is there any scenario where a key would be in more than one index partition? if not, then a length of 1 should be sufficient. if the reason for len(p.Partition) is to not block the insertion into the chan:

further down i think it would make more sense to only wait for the first result in the chan and then not wait for the other routines to finish, otherwise we might end up waiting for a search to complete even though we already have the result. Unfortunately I can't see how to do that with `errgroup.WithContext`.

File: idx/memory/partitioned_idx.go (hunk 3 above)
Comment (replay) [inline idx/memory/partitioned_idx.go]: is there any scenario where a key would be in more than one index partition? if not, then a length of 1 should be sufficient. if the reason for len(p.Partition) is to not block the insertion into the chan:

further down i think it would make more sense to only wait for the first result in the chan and then not wait for the other routines to finish, otherwise we might end up waiting for a search to complete even though we already have the result. Unfortunately I can't see how to do that with errgroup.WithContext unless you .Wait() in a separate routine

File: idx/memory/partitioned_idx.go [hunk 5]
```diff
@@ -0,0 +1,536 @@
...
+	var i int
+	for _, m := range p.Partition {
```
Comment (replay) [inline idx/memory/partitioned_idx.go]: instead of `_` this could just be `i` and then the var wouldn't need to be initialized above nor increased further down, because its actual value seems to make no difference in this use case, only its uniqueness

File: idx/memory/partitioned_idx.go [hunk 6]
```diff
@@ -0,0 +1,536 @@
...
+// If the pattern matches a branch node, then
+// all leaf nodes on that branch are deleted. So if the pattern is
+// "*", all items in the index are deleted.
+// It returns a copy of all of the Archives deleted.
+func (p *PartitionedMemoryIdx) Delete(orgId uint32, pattern string) ([]idx.Archive, error) {
+	g, _ := errgroup.WithContext(context.Background())
+	result := make([][]idx.Archive, len(p.Partition))
+	var i int
+	for _, m := range p.Partition {
+		pos, m := i, m
```
Comment (replay) [inline idx/memory/partitioned_idx.go]: same as above^^

File: idx/memory/partitioned_idx.go [hunk 7]
```diff
@@ -0,0 +1,536 @@
...
+// * orgId describes the org to search in (public data in orgIdPublic is automatically included)
+// * pattern is handled like graphite does. see https://graphite.readthedocs.io/en/latest/render_api.html#paths-and-wildcards
+// * from is a unix timestamp. series not updated since then are excluded.
+func (p *PartitionedMemoryIdx) Find(orgId uint32, pattern string, from int64) ([]idx.Node, error) {
+	g, _ := errgroup.WithContext(context.Background())
+	resultChan := make(chan []idx.Node)
+	for _, m := range p.Partition {
+		m := m
+		g.Go(func() error {
+			//fmt.Printf("running find on %d\n", pos)
```
Comment (replay) [inline idx/memory/partitioned_idx.go]: commented printfs

File: idx/memory/partitioned_idx.go [hunk 8]
```diff
@@ -0,0 +1,536 @@
...
+			found, err := m.Find(orgId, pattern, from)
+			if err != nil {
+				return err
+			}
+			resultChan <- found
+			//fmt.Printf("completed find on %d\n", pos)
```
Comment (replay) [inline idx/memory/partitioned_idx.go]: same^^

File: idx/memory/partitioned_idx.go [hunk 9]
```diff
@@ -0,0 +1,536 @@
...
+		})
+	}
+
+	// consume from the ResultChan and merge the results, de-duping branches
+	byPath := make(map[string]*idx.Node)
+	done := make(chan struct{})
+	go func() {
+		for found := range resultChan {
+			for _, node := range found {
+				if _, ok := byPath[node.Path]; !ok {
```
Comment (replay) [inline idx/memory/partitioned_idx.go]: if instead of `_` this would assign the value to a variable then this variable could be used in the `if` / `else` block instead of doing `byPath[node.Path]` many times. if the result set is large we can save a bunch of lookups in large `byPath` maps by doing that.

File: idx/memory/partitioned_idx.go (hunk 5 above)
Comment (woodsaj) [reply idx/memory/partitioned_idx.go]: `p.Partition` is a map, not a slice.

File: idx/memory/partitioned_idx.go (hunk 6 above)
Comment (woodsaj) [reply idx/memory/partitioned_idx.go]: as above

File: idx/memory/partitioned_idx.go (hunk 3 above)
Comment (woodsaj) [reply idx/memory/partitioned_idx.go]: a specific key should only ever be in one partition. However, we need to make the resultChan big enough for all goroutines to write to it just in-case.  Otherwise we will end up with a deadlock.

File: idx/memory/partitioned_idx.go (hunk 5 above)
Comment (replay) [reply idx/memory/partitioned_idx.go]: i know, so `i` will be the key of the map, but the resulting return value from the function should remain the same

File: idx/memory/partitioned_idx.go (hunk 3 above)
Comment (replay) [reply idx/memory/partitioned_idx.go]: What I meant to describe was more like this:
If we know there will never be more than 1 result, and we have 1 result, then we can stop waiting for more results.
```
func (p *PartitionedMemoryIdx) Get(key schema.MKey) (idx.Archive, bool) {
	g, _ := errgroup.WithContext(context.Background())
	resultChan := make(chan *idx.Archive, len(p.Partition))
	for _, m := range p.Partition {
		m := m
		g.Go(func() error {
			if a, ok := m.Get(key); ok {
				resultChan <- &a
			}
			return nil
		})
	}

	var err atomic.Value
	go func() {
		if e := g.Wait(); e != nil {
			err.Store(e)
		}
		close(resultChan)
	}()

	result, ok := <-resultChan
	if !ok {
		e := err.Load()
		if e != nil {
			log.Errorf("memory-idx: failed to get Archive with key %v. %s", key, e)
		}
		return idx.Archive{}, false
	}

	return *result, true
}
```

File: idx/memory/partitioned_idx.go (hunk 5 above)
Comment (woodsaj) [reply idx/memory/partitioned_idx.go]: i is used as the index into the result slice.  You cant use the partition number for this.  Only the first shard would have partition numbers 0,1,2,3,4,5,6,7.  then next shard would have partitions 8,9,10,11,12,13,14,15

**Comment by woodsaj** [comment]
thanks for the review @replay 
I think this should be good to merge. The change is pretty safe given that by default the unpartitionedIdx will be used.

relevant benchmark
```
goos: linux
goarch: amd64
pkg: github.com/grafana/metrictank/idx/memory
BenchmarkConcurrentInsertFind/partitioned-8                  10000            363256 ns/op          524709 B/op      11788 allocs/op
BenchmarkConcurrentInsertFind/unPartitioned-8               3000            333891 ns/op          484942 B/op      10115 allocs/op
PASS
ok      github.com/grafana/metrictank/idx/memory        12.061s
```

For most benchmarks the the PartitionedMemoryIdx performs worse than the unpartitioned version. This is all due to the results from searching all partitions needing to be merged, which results in lots of allocations.

The next step is to continue exposing "partitions" throughout the code base. The end goal is to have all partitions handled independently and allow peers to query specific partitions.  This will allow nodes to respond with data for some partitions while others are blocked.  The speculative execution can then try alternate peers for partitions that are slow.

File: idx/memory/partitioned_idx.go (hunk 5 above)
Comment (replay) [reply idx/memory/partitioned_idx.go]: I'm starting to feel like I must be missing something :joy:. Currently the key in `result` is an increasing int starting from `0`, where each value is unique for one partition. If you'd use the keys of `p.Partition` instead of that increasing int, then the keys in `result` wouldn't necessarily be beginning from `0` anymore because some MT consume the higher partitions (as you're describing). But my point is that the value of the keys in `result` is irrelevant because it is not used anywhere, so it doesn't matter if those keys are starting at `0` or some higher value. Honestly this change really isn't very important, it's only going to save the allocation of one int.

**Comment by replay** [comment]
@woodsaj sounds good. I'd just like to ask first what you think about my above suggestion, because i feel like this could help a lot for multi tenant setups with many tenants:

> if we already partition by partition id, should we also partition by org id? in the case of multi tenant setups this would keep the defById map smaller and it seems like it should be easy to now just make the key something like (orgId * 128) + partition assuming the max number of partitions is 128.

Instead of keeping it a single map of index partitions we could also add a second level like `Partitions[orgId][partitionId]*MemoryIdx`

File: idx/memory/partitioned_idx.go (hunk 5 above)
Comment (woodsaj) [reply idx/memory/partitioned_idx.go]: @replay https://play.golang.org/p/Gu_xXmkNRuN

**Comment by woodsaj** [comment]
> because i feel like this could help a lot for multi tenant setups with many tenants:

it could, but the majority of MT clusters are single tenant. 

It might still make sense to do, but it is not something that should be included in this PR

**Comment by Dieterbe** [comment]
@woodsaj at this point, would you recommend enabling the partitioning ? in which circumstances?

File: idx/cassandra/cassandra.go [hunk 10]
```diff
@@ -315,8 +315,12 @@ func (c *CasIdx) updateCassandra(now uint32, inMemory bool, archive idx.Archive,
 func (c *CasIdx) rebuildIndex() {
 	log.Info("cassandra-idx: Rebuilding Memory Index from metricDefinitions in Cassandra")
 	pre := time.Now()
-	defs := c.LoadPartitions(cluster.Manager.GetPartitions(), nil, pre)
-	num := c.MemoryIdx.Load(defs)
+	var defs []schema.MetricDefinition
+	var num int
+	for _, partition := range cluster.Manager.GetPartitions() {
+		defs = c.LoadPartitions([]int32{partition}, defs[:0], pre)
```
Comment (Dieterbe) [inline idx/cassandra/cassandra.go]: @woodsaj isn't this a bug? subsequent calls of this seem to overwrite previously loaded defs?

File: idx/memory/memory.go [hunk 11]
```diff
@@ -407,13 +431,12 @@ func (m *MemoryIdx) Load(defs []schema.MetricDefinition) int {
...
 }
 
-func (m *MemoryIdx) add(def *schema.MetricDefinition) idx.Archive {
+func (m *UnpartitionedMemoryIdx) add(def *schema.MetricDefinition) idx.Archive {
 	path := def.NameWithTags()
-
 	schemaId, _ := mdata.MatchSchema(path, def.Interval)
 	aggId, _ := mdata.MatchAgg(path)
 	irId, _ := IndexRules.Match(path)
-	sort.Strings(def.Tags)
```
Comment (Dieterbe) [inline idx/memory/memory.go]: why are we removing this? this looks like a change that requires special attention, and not slip in via a commit named "add new PartitionedMemoryIdx".
ideally we don't make changes in this PR that are unrelated to the partitioning.

File: idx/cassandra/cassandra.go (hunk 10 above)
Comment (woodsaj) [reply idx/cassandra/cassandra.go]: no.

File: idx/memory/memory.go (hunk 11 above)
Comment (woodsaj) [reply idx/memory/memory.go]: `def.NameWithTags()` already sorts the tags.
